### PR TITLE
[Previous URL] Expand to embedded protocol and filesystem importers

### PR DESCRIPTION
**Pull request by nex3**
This updates the embedded protocol to support the proposal and adds
support for passing the previous URL to filesystem importers.

### CODE REVIEW

**Comment by nex3** [comment]
@ntkme I'm interested in your feedback on this as well.

File: proposal/prev-url.d.ts.md [hunk 1]
```diff
@@ -209,5 +253,67 @@ nonCanonicalScheme?: string | string[];
...
+## Embedded Protocol
+
+### `Importer`
+
+#### `non_canonical_scheme`
+
+The set of URL schemes that are considered *non-canonical* for this importer.
+This must be empty unless `importer.importer_id` is set.
+
+If any element of this contains a character other than a lowercase ASCII letter,
```
Comment (Goodwine) [inline proposal/prev-url.d.ts.md:268]: nitpicking: empty string should also throw an error

File: proposal/prev-url.d.ts.md [hunk 2]
```diff
@@ -209,5 +253,67 @@ nonCanonicalScheme?: string | string[];
...
+```
+
+## Embedded Protocol
+
+### `Importer`
+
+#### `non_canonical_scheme`
+
+The set of URL schemes that are considered *non-canonical* for this importer.
+This must be empty unless `importer.importer_id` is set.
```
Comment (Goodwine) [inline proposal/prev-url.d.ts.md:266]: Can it be empty when `importer.importer_id` is set?

File: proposal/prev-url.d.ts.md
Comment (Goodwine) [inline proposal/prev-url.d.ts.md:1]: I realize it's not part of this PR, but I wonder if the TS type for `nonCanonicalScheme` should be:

```ts
  nonCanonicalScheme?: string | [string, ...string[]];
```

Just to capture that if an array is provided it will have at least 1 element

**Review by Goodwine — COMMENTED** [review]
Is the idea for this proposal as a whole (ie not just this PR) that custom importers would be able to use custom schemas to emulate NodeJS behavior of using the closest `node_module`?

e.g. from the Issue, for a file: `a/node_modules/b/node_modules`, URLs loaded from `/a/` could have `schema1://` and files loaded from `/b/` could have `schema2://`? 

Otherwise, I'm slightly confused how the non-canonical schemes come into play after reading the summary 🤣😭 (the relative loads do make sense)

File: proposal/prev-url.d.ts.md
Comment (nex3) [reply proposal/prev-url.d.ts.md:1]: We do have well-defined semantics for an empty array (all schemes are considered canonical, just like `undefined`) so I don't think that adds much value. For example, consider a user who might want to assemble this array programmatically—it may indeed be more convenient for them to allow it to be empty.

File: proposal/prev-url.d.ts.md (hunk 2 above)
Comment (nex3) [reply proposal/prev-url.d.ts.md:266]: Yes, it just means that no URL schemes are considered non-canonical.

**Review by nex3 — COMMENTED** [review]
One use-case for non-canonical schemes is indeed to allow for https://github.com/sass/sass/issues/2739, where we have a `pkg:` URL scheme that can resolve differently based on the nearest `node_modules` location.

**Review by Goodwine — APPROVED** [review]
